docs/dependency-injection.md

```diff
@@ -0,0 +1,190 @@
+
+Blazor's implementation of `System.IServiceProvider` obtains its services from an underlying [Microsoft.Extensions.DependencyInjection.IServiceCollection](https://docs.microsoft.com/dotnet/api/microsoft.extensions.dependencyinjection.iservicecollection).
+
+## Add services to DI
+
+After creating a new Blazor app, examine the `Main` method in *Program.cs*:
+
+```csharp
+static void Main(string[] args)
+{
+    var serviceProvider = new BrowserServiceProvider(configure =>
+    {
+        // Add any custom services here
+    });
+
+    new BrowserRenderer(serviceProvider).AddComponent<App>("app");
+}
+```
+
+`BrowserServiceProvider` receives an action with which you can add your app services to Blazor's DI. `configure` is referencing the underlying `IServiceCollection`, which is a list of service descriptor objects ([Microsoft.Extensions.DependencyInjection.ServiceDescriptor](https://docs.microsoft.com/dotnet/api/microsoft.extensions.dependencyinjection.servicedescriptor)). Services are added by providing service descriptors to the service collection. The following is a code sample demonstrating the concept:
+
+```csharp
+static void Main(string[] args)
+{
+    var serviceProvider = new BrowserServiceProvider(configure =>
+    {
+        configure.Add(ServiceDescriptor.Singleton<IDataAccess, DataAccess>());
+    });
+
+    new BrowserRenderer(serviceProvider).AddComponent<App>("app");
+}
+```
+
+`ServiceDescriptor` offers various overloads of three different functions that can be used to add services to Blazor's DI:
+
+| Method      | Description |
+| ----------- | ----------- |
+| `Singleton` | Blazor creates a *single instance* of your service. All components requiring this service receive a reference to this instance. |
+| `Transient` | Whenever a component requires this service, it receives a *new instance* of the service. |
+| `Scoped`    | Blazor doesn't currently have the concept of DI scopes. `Scoped` behaves like `Singleton`. Therefore, prefer `Singleton` and avoid `Scoped`. |
+
+## System services
+
+Blazor provides default services that are automatically added to the service collection of a Blazor app. The following table shows a list of the default services currently provided by Blazor.
+
+| Method       | Description |
+| ------------ | ----------- |
+| `IUriHelper` | Helpers for working with URIs and navigation state (singleton). |
+| `HttpClient` | Provides methods for sending HTTP requests and receiving HTTP responses from a resource identified by a URI (singleton). Note that this instance of [System.Net.Http.HttpClient](https://docs.microsoft.com/dotnet/api/system.net.http.httpclient) is using the browser for handling the HTTP traffic in the background. Its [BaseAddress](https://docs.microsoft.com/dotnet/api/system.net.http.httpclient.baseaddress) is automatically set to the base URI prefix of the Blazor app. |
```

Comment:
In a couple of places it says "is using" or "is referencing". Maybe it's a language difference, but a more common phrasing would be simply "uses" or "references". Not a big deal though!